pkg/volume/azure_dd/common.go

```diff
@@ -0,0 +1,322 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure_dd
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"regexp"
+	"strconv"
+	STRINGS "strings"
+
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/azure"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/util/strings"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+const (
+	default_fstype               = "ext4"
+	default_storage_account_type = "standard_lrs"
+)
+
+type dataDisk struct {
+	volume.MetricsProvider
+	volumeName string
+	diskName   string
+	podUID     types.UID
+}
+
+var supportedCachingModes = sets.NewString(
+	string(api.AzureDataDiskCachingNone),
+	string(api.AzureDataDiskCachingReadOnly),
+	string(api.AzureDataDiskCachingReadWrite),
+)
+
+var supportedDiskKinds = sets.NewString(
+	string(api.AzureSharedBlobDisk),
+	string(api.AzureDedicatedBlobDisk),
+	string(api.AzureManagedDisk),
+)
+
+var supportedStorageAccountTypes = sets.NewString("premium_lrs", "standard_lrs")
+
+func getPath(uid types.UID, volName string, host volume.VolumeHost) string {
+	return host.GetPodVolumeDir(uid, strings.EscapeQualifiedNameForDisk(azureDataDiskPluginName), volName)
+}
+
+// creates a unique path for disks (even if they share the same *.vhd name)
+func makeGlobalPDPath(host volume.VolumeHost, diskUri string, isManaged bool) (string, error) {
+	diskUri = STRINGS.ToLower(diskUri) // always lower uri because users may enter it in caps.
+	uniqueDiskNameTemplate := "%s%s"
+	az, _ := host.GetCloudProvider().(*azure.Cloud)
```

Comment:
by the time this function is called, validation has already occurred at the time of Init() call on the plugin